Read results and update skipfile

katieworton opened the pull request:
Redraft of skipfile reading results

To do before review:
- [x] Get a generated list of boards that would be counted as "all"
- [x] Remove/align the `lib.sh` and `squad-register-skipfile-callback.sh` with the original code from LKFT pipeline https://gitlab.com/Linaro/lkft/pipelines/common/ so these don't have to be maintained in here 
- [x] General tidy up of `read-skipfile-results` and resolving of outstanding TODOs

katieworton committed:
read-skipfile-results: Create script for reading ltp skipfile test results

Create script for reading skipfile results as part of the skipfile
pipeline.

Tux plan files generated by squad-create-skipfile-reproducers are
submitted by the Gitlab CI pipeline, which logs the SQUAD build names
for the results into a text file. The read-skipfile-results script then
loops through the list of SQUAD builds and reads the results of the
skipfile testing (counting the non-hang tests in a dataframe). For each
test, the number of non-hang runs is compared to the total number of
runs, to determine which skipfile entries are safe to remove. Then, the
script will update the skipfile in the test-definitions GitHub repo and
optionally push these updates as PRs to the test-definitions repo. The
results dataframes are also stored to csv files in test-definitions for
inspection (not tracked by git).

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
squadutilslib: return git describe

Add the git describe to the values returned by get_reproducer. This way
the value can be logged and later retrieved in the skipfile pipeline.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

roxell commented on run-and-push-skipfile-results.sh:
```diff
@@ -6,11 +6,19 @@ PROJECT_NAME=$2
-        echo $BUILD >>builds_for_skipfile_runs.txt
+        if test -f "$PLAN"; then
+            echo "$PLAN"
+            tuxsuite plan $PLAN --json-out $PLAN.json --no-wait
```
QA_TEAM=~katie.worton
QA_PROJECT=skipfile-testing
tuxsuite plan $PLAN --json-out $PLAN.json --no-wait --callback ${QA_SERVER}/api/fetchjob/${QA_TEAM}/${QA_PROJECT}/${git_desc}/env/tuxsuite.com

katieworton replied:
Contents of script moved to pipeline config

katieworton committed:
squad-create-skipfile-reproducers: add multi runs

Add support for running skipfile tests multiple times.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
squadutilslib: fix parameters key

Correct the output of script when parameters key is in the tuxsuite
reproducer and the parameter value contains a `=` assignment.

Previously these would be incorrectly translated to the format:
    parameters:
      cpu.lpa2=off

Now they are correctly translated to:
    parameters:
      cpu.lpa2:
        off

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
run-and-push-skipfile-results: delete script

Delete the run-and-push-skipfile-results script as this is now
maintained within the skipfile pipeline repository.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
squadutilslib: add build wait function

Add function that takes a list of build names, and waits until all of
the builds from this list are marked as finished.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton commented:
@roxell if you could review this when you have the opportunity that would be great :raised_hands:

roxell commented on read-skipfile-results:
```diff
@@ -23,7 +23,7 @@ from squad_client.core.api import SquadApi
 from squad_client.utils import first, getid
 
-from squadutilslib import get_file
+from squadutilslib import get_file, wait_for_builds
```
This can be two commits one in the lib and one where you use it.

roxell commented on read-skipfile-results:
```diff
@@ -0,0 +1,433 @@
+from pathlib import Path
+from re import findall, sub
+
+import pandas as pd
```
the requirements.txt file should be updated with pandas in this commit.

katieworton replied:
I have updated the requirements.txt to reflect the requirements of the scripts

roxell commented on read-skipfile-results:
```diff
@@ -0,0 +1,433 @@
+        "--github-push",
+        required=False,
+        default=True,
+        type=int,
```
should not be int.
Default should probably be False so you don't push... and if you push you just add --github-push or --push-to-github or ...

roxell commented on read-skipfile-results:
```diff
@@ -0,0 +1,433 @@
+    )
+
+    parser.add_argument(
+        "--project_name",
```
rename it to project-name.

roxell commented on read-skipfile-results:
```diff
@@ -0,0 +1,433 @@
+    parser.add_argument(
+        "--username",
+        required=False,
+        default="katieworton",
```
no default please

roxell commented on read-skipfile-results:
```diff
@@ -0,0 +1,433 @@
+#!/usr/bin/env python3
```
add more to the changelog please.

also add an example how to run it.
what you have to run first to get the builds_for_...txt and metadata_list.csv

katieworton replied:
I updated the README to reflect requirements and usage. Have added more to commit message to explain purpose. Let me know if this is appropriate.

roxell commented on read-skipfile-results:
```diff
@@ -0,0 +1,433 @@
+                            + f"success count: {test_dict[tests_string].loc[device, build_name]}",
+                        )
+
+                    except KeyError as e:
```
Can we find something else than KeyError?

roxell commented on read-skipfile-results:
```diff
@@ -0,0 +1,433 @@
+                    reproducer = open("results", "r")
+                    results_json = load(reproducer)
+                    try:
+                        # Get names of tests run and split it into a list
```
This shouldn't be needed if we push the command_name as just 'fork13' instead of
'cd /opt/ltp && ./runltp -s fork13'

roxell commented on squadutilslib.py:
```diff
@@ -268,6 +268,13 @@ def tuxtest_to_tuxplan_entry(tuxsuite_test):
                 else:
                     dict_entry[key] = dict()
                     dict_entry[key][test] = int(timeout)
+            elif key == "parameters":
```
please update the changelog with what was the problem

roxell commented:
patch "run-and-push-skipfile-results: delete script" maybe say that we remove it since we push the resutls when submitting to tuxsuite with the callback command.

katieworton committed:
squadutilslib: bug fix for command-name parameter

Bug fix for the command-name parameter, as previously this was being
passed in incorrectly causing it to be silently dropped by tuxsuite.

Before the fix, it was being incorrectly passed in as `--command-name
<command-name>` but after this fix it will now be correctly passed in as
`--parameters command-name=<command-name>`. This then allows the command
name to be read from SQUAD.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
squadutilslib: Allow reproducer command-name customisation

Add a command_name parameter to the create_custom_reproducer function
that will set the command-name for a reproducer to something other than
the contents of custom_commands. If the new command_name parameter is
not provided, the command-name will default to the contents of
custom_commands.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
squadutilslib: Add function to create command-name

Create a function called generate_command_name_from_list which generates
a valid command-name for custom commands, with the list entries
alphabetised. This will be used for creating valid command-name values
for skipfile test lists and then comparing test lists with command-name
strings.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
squad-create-skipfile-reproducers: set valid command-name

Generate a valid custom command name using the
generate_command_name_from_list function, then set this as the
command-name in tuxsuite skipfile reproducers.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
squadutilslib: add build_name to reproducer info

Add build_name to the information return by the get_reproducer function
to make it easier to reproduce results of the skipfile pipeline.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
squad-create-skipfile-reproducer: update to use build_name

Update script to incorporate build name into the reproducer info

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
squad-create-skipfile-reproducers: Update to log git desc

Update the squad-create-skipfile-reproducers script to align with
changes to get_reproducer which now returns the reproducer and git
describe.

Log the squad build name, project, device and git describe to a csv file
so they can be retrieved later in the skipfile pipeline.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
squad-create-reproducer: Update get_reproducer call

Update the returned items from get_reproducer since this now returns
both the reproducer and the git describe.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
squad-create-reproducer: Update get_reproducer call with build_name

Update the get_reproducer call since get_reproducer now returns the
reproducer, git describe and build_name

Signed-off-by: Katie Worton <katie.worton@linaro.org>
